common/src/main/java/org/apache/celeborn/common/network/server/CreditStreamManager.java

```diff
@@ -0,0 +1,300 @@
+
+    logger.debug("Register stream streamId: {}, fileInfo: {}", streamId, fileInfo);
+
+    return streamId;
+  }
+
+  private void addCredit(MapDataPartition mapDataPartition, int numCredit, long streamId) {
+    logger.debug("streamId: {}, add credit: {}", streamId, numCredit);
+    try {
+      if (mapDataPartition != null && numCredit > 0) {
+        mapDataPartition.addReaderCredit(numCredit, streamId);
+      }
+    } catch (Throwable e) {
+      logger.error("streamId: {}, add credit end: {}", streamId, numCredit);
+    }
+  }
+
+  public void addCredit(int numCredit, long streamId) {
+    MapDataPartition mapDataPartition = streams.get(streamId).getMapDataPartition();
+    addCredit(mapDataPartition, numCredit, streamId);
+  }
+
+  public void connectionTerminated(Channel channel) {
+    for (Map.Entry<Long, StreamState> entry : streams.entrySet()) {
+      if (entry.getValue().getAssociatedChannel() == channel) {
+        logger.info("connection closed, clean streamId: {}", entry.getKey());
+        recycleStream(entry.getKey());
+      }
+    }
+  }
+
+  public void notifyStreamEndByClient(long streamId) {
+    recycleStream(streamId);
+  }
+
+  public void recycleStream(long streamId) {
+    recycleStreamIds.add(new DelayedStreamId(streamId));
+    startRecycleThread(); // lazy start thread
+  }
+
+  @VisibleForTesting
+  public int numStreamStates() {
+    return streams.size();
+  }
+
+  @VisibleForTesting
+  public int numRecycleStreams() {
+    return recycleStreamIds.size();
+  }
+
+  @VisibleForTesting
+  public ConcurrentHashMap<Long, StreamState> getStreams() {
+    return streams;
+  }
+
+  private void startRecycleThread() {
+    if (recycleThread == null) {
+      synchronized (lock) {
+        if (recycleThread == null) {
+          recycleThread =
+              new Thread(
+                  () -> {
+                    while (true) {
+                      try {
+                        DelayedStreamId delayedStreamId = recycleStreamIds.take();
+                        cleanResource(delayedStreamId.streamId);
+                      } catch (Throwable e) {
+                        logger.warn(e.getMessage(), e);
+                      }
+                    }
+                  },
+                  "recycle-thread");
+          recycleThread.setDaemon(true);
+          recycleThread.start();
+
+          logger.info("start stream recycle thread");
+        }
+      }
+    }
+  }
+
+  public void cleanResource(Long streamId) {
+    logger.debug("received clean stream: {}", streamId);
+    if (streams.containsKey(streamId)) {
+      MapDataPartition mapDataPartition = streams.get(streamId).getMapDataPartition();
+      if (mapDataPartition != null) {
+        if (mapDataPartition.releaseStream(streamId)) {
+          StreamState state = streams.remove(streamId);
+          MemoryManager.instance()
+              .removeReadBufferTargetChangeListener(state.getMapDataPartition());
```

Comment:
Should not remove MapPartition because there might be other streams. I think you got it wrong with MapDataPartition and stream reader here, please re-check.